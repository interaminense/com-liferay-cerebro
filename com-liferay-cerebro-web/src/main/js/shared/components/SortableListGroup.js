import autobind from 'autobind-decorator';
import Component, {Config} from 'metal-jsx';

import Icon from './Icon';
import ListGroup from './list-group';
import SortableList from '../util/SortableList';

class SortableListGroupItem extends Component {
	@autobind
	handleRemove() {
		const {id, onRemove} = this.props;

		onRemove(id);
	}

	render() {
		const {name} = this.props;

		return (
			<ListGroup.Item flex>
				<ListGroup.ItemField elementClasses="handle">
					<Icon symbol="bars" />
				</ListGroup.ItemField>

				<ListGroup.ItemField expand>{name}</ListGroup.ItemField>

				<ListGroup.ItemField>
					<Icon
						elementClasses="remove"
						onClick={this.handleRemove}
						symbol="times-circle"
					/>
				</ListGroup.ItemField>
			</ListGroup.Item>
		);
	}
}

SortableListGroupItem.PROPS = {
	id: Config.oneOfType([Config.number(), Config.string()]),
	name: Config.string(),
	onRemove: Config.func()
};

class SortableListGroup extends Component {
	attached() {
		this._sortableList = new SortableList({
			axis: 'y',
			container: this.element,
			items: this.props.items,
			onItemsChange: this.props.onChange,
			sources: '.list-group-item',
			targets: '.list-group-item'
		});
	}

	@autobind
	handleRemove(id) {
		const {idKey, items, onChange} = this.props;

		onChange(items.filter(item => item[idKey] !== id));
	}

	syncItems(items) {
		if (this._sortableList) {
			this._sortableList.items = items;
		}
	}

	render() {
		const {idKey, nameKey} = this.props;

		return (
			<ListGroup elementClasses="sortable-list-group-root">
				{this.props.items.map(item => (
					<SortableListGroupItem
						id={item[idKey]}
						key={item[idKey]}
						name={item[nameKey]}
						onRemove={this.handleRemove}
					/>
				))}
			</ListGroup>
		);
	}
}

SortableListGroup.PROPS = {
	idKey: Config.string().value('id'),
	items: Config.array(),
	nameKey: Config.string().value('name'),
	onChange: Config.func()
};

export default SortableListGroup;