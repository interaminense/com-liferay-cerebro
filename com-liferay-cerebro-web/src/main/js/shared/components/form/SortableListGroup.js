import Component, {Config} from './Component';
import {identity, uniqueId} from 'lodash';

import SortableListGroup from 'shared/components/SortableListGroup';

class FormSortableList extends Component {
	created() {
		this._id = uniqueId('FormSortableList');

		const {items, valueFn} = this.props;

		if (items) {
			this.state.value = valueFn(items);
		}
	}

	syncItems(newVal) {
		this.state.value = this.props.valueFn(newVal);
	}

	render() {
		const {items} = this.props;

		return <SortableListGroup {...this.otherProps()} items={items} />;
	}
}

FormSortableList.PROPS = {
	items: Config.array().value([]),
	valueFn: Config.func().value(identity)
};

FormSortableList.STATE = {
	value: Config.array().value([])
};

export default FormSortableList;