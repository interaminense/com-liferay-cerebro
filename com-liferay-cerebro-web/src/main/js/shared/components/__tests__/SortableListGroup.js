import SortableListGroup from '../SortableListGroup';

describe('SortableListGroup', () => {
	let component;

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render', () => {
		component = new SortableListGroup({
			items: [{id: 1, name: 'foo'}, {id: 2, name: 'bar'}]
		});

		expect(component).toMatchSnapshot();
	});
});