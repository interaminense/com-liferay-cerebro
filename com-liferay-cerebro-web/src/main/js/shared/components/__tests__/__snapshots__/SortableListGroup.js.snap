// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortableListGroup should render 1`] = `
<ul class="list-group list-group-root sortable-list-group-root">
  <li class="list-group-item list-group-item-flex">
    <div class="flex-col handle">
      <svg class="icon-root lexicon-icon lexicon-icon-bars">
        <use xlink:href="/images/lexicon/icons.svg#bars"></use>
      </svg>
    </div>
    <div class="flex-col flex-col-expand">foo</div>
    <div class="flex-col">
      <svg class="icon-root lexicon-icon lexicon-icon-times-circle remove">
        <use xlink:href="/images/lexicon/icons.svg#times-circle"></use>
      </svg>
    </div>
  </li>
  <li class="list-group-item list-group-item-flex">
    <div class="flex-col handle">
      <svg class="icon-root lexicon-icon lexicon-icon-bars">
        <use xlink:href="/images/lexicon/icons.svg#bars"></use>
      </svg>
    </div>
    <div class="flex-col flex-col-expand">bar</div>
    <div class="flex-col">
      <svg class="icon-root lexicon-icon lexicon-icon-times-circle remove">
        <use xlink:href="/images/lexicon/icons.svg#times-circle"></use>
      </svg>
    </div>
  </li>
</ul>
`;
